faculty/clients/template.py

```diff
@@ -42,4 +45,310 @@ def publish_new(self, project_id, template, source_directory):
+
+
+ApiValidationFailure = namedtuple(
+    "ApiValidationFailure",
+    ["subdomain_conflicts", "name_conflicts", "invalid_working_dirs"],
+)
+
+
+class EnvironmentValidationFailureSchema(BaseSchema):
+    name_conflicts = fields.List(
+        fields.String(), data_key="nameConflicts", required=True
+    )
+    invalid_names = fields.List(
+        fields.String(), data_key="invalidNames", required=True
+    )
+
+    @post_load
+    def make_failure(self, data, **kwargs):
+        return EnvironmentValidationFailure(**data)
+
+
+EnvironmentValidationFailure = namedtuple(
+    "EnvironmentValidationFailure", ["name_conflicts", "invalid_names"]
+)
+
+
+class JobValidationFailureSchema(BaseSchema):
+    name_conflicts = fields.List(
+        fields.String(), data_key="nameConflicts", required=True
+    )
+    invalid_working_dirs = fields.List(
+        fields.String(), data_key="invalidWorkingDirs", required=True
+    )
+    invalid_names = fields.List(
+        fields.String(), data_key="invalidNames", required=True
+    )
+
+    @post_load
+    def make_failure(self, data, **kwargs):
+        return JobValidationFailure(**data)
+
+
+JobValidationFailure = namedtuple(
+    "JobValidationFailure",
+    ["name_conflicts", "invalid_working_dirs", "invalid_names"],
+)
+
+
+class WorkspaceValidationFailureSchema(BaseSchema):
+    name_conflicts = fields.List(
+        fields.String(), data_key="nameConflicts", required=True
+    )
+
+    @post_load
+    def make_failure(self, data, **kwargs):
+        return WorkspaceValidationFailure(**data)
+
+
+WorkspaceValidationFailure = namedtuple(
+    "WorkspaceValidationFailure", ["name_conflicts"]
+)
+
+
+class ResourceValidationFailuresSchema(BaseSchema):
+    apps = fields.Nested(AppValidationFailureSchema(), required=True)
+    apis = fields.Nested(ApiValidationFailureSchema(), required=True)
+    environments = fields.Nested(
+        EnvironmentValidationFailureSchema(), required=True
+    )
+    jobs = fields.Nested(JobValidationFailureSchema(), required=True)
+    workspace = fields.Nested(
+        WorkspaceValidationFailureSchema(), required=True
+    )
+
+
+class ResourceValidationFailureResponseSchema(BaseSchema):
+    errors = fields.Nested(ResourceValidationFailuresSchema(), required=True)
+
+    @post_load
+    def make_failure(self, data, **kwargs):
+        return ResourceValidationFailure.from_errors(**data)
+
+
+class ResourceValidationFailure(TemplateException):
+    def __init__(self, apps, apis, environments, jobs, workspace):
+        self.apps = apps
+        self.apis = apis
+        self.environments = environments
+        self.jobs = jobs
+        self.workspace = workspace
+
+    @staticmethod
+    def from_errors(errors):
+        return ResourceValidationFailure(
+            apps=errors["apps"],
+            apis=errors["apis"],
+            environments=errors["environments"],
+            jobs=errors["jobs"],
+            workspace=errors["workspace"],
+        )
```

Comment:
Weak suggestion avoiding the need for the `classmethod`:

```suggestion
class ResourceValidationFailureResponseSchema(BaseSchema):
    errors = fields.Nested(ResourceValidationFailuresSchema(), required=True)

    @post_load
    def make_failure(self, data, **kwargs):
        errors = data["errors"]
        return ResourceValidationFailure.from_errors(**errors)


class ResourceValidationFailure(TemplateException):
    def __init__(self, apps, apis, environments, jobs, workspace):
        self.apps = apps
        self.apis = apis
        self.environments = environments
        self.jobs = jobs
        self.workspace = workspace
```

Reply:
I think you meant to call the constructor on line 313? Not `from_errors`(because you just removed it). 

If so, then yes, that makes sense, thanks.